### ISSUE TITLE
Better outlier filter for trainset

### ISSUE THREAD

**Issue by user1823** [open-spaced-repetition/fsrs-rs#121]
> As I mentioned in https://github.com/open-spaced-repetition/fsrs-rs/issues/88#issuecomment-1738765850, the current outlier filter is too aggressive (it removes too many reviews). This is fine for the pretrain function (because the pretrain method is fragile). However, the train function should have access to more reviews.
> 
> I agree that some sort of outlier filter is required for the trainset also. But, I think that the same outlier filter should not be used for both.

_Originally posted by @user1823 in https://github.com/open-spaced-repetition/fsrs-rs/issues/119#issuecomment-1820856409_

> Taking the example of my collection (data: [stability_for_pretrain.zip](https://github.com/open-spaced-repetition/fsrs-rs/files/13427569/stability_for_pretrain.zip))
> 
> - For first_rating = 1, I think that it is reasonable to keep delta_t = 1 and delta_t = 2 (in my collection) in the trainset.
> - For first_rating = 2, the data in my collection is insufficient. So, I will not discuss about it.
> - For first_rating = 3, the initial stability is calculated as about 14. So, it doesn't make sense to ONLY keep the data with delta_t < 5 in the trainset.
> - For first_rating = 4, there is no major problem in how the outlier filter is working with my data.
> 

_Originally posted by @user1823 in https://github.com/open-spaced-repetition/fsrs-rs/issues/119#issuecomment-1821064696_

> > If a card is reviewed too early or too late, the first response will be unreliable. 
> 
> So, what if create another condition based on the ratio (or something similar) of the delta_t and stability and then remove the reviews of cards that fulfill BOTH the conditions. (By both, I mean the ratio condition and the pretrain outlier condition.)
> 
> Note: This suggestion is for trainset. I recommend keeping the pretrain filter unchanged.

_Originally posted by @user1823 in https://github.com/open-spaced-repetition/fsrs-rs/issues/119#issuecomment-1821219828_

**Comment by user1823** [comment]
@L-M-Sherlock, what is opinion on the suggestion?

@Expertium, you may also be interested.

**Comment by Expertium** [comment]
@user1823 can you elaborate how exactly the new condition should work?

**Comment by user1823** [comment]
I was thinking of using `stability × 3` and `stability / 3` as the maximum and minimum values of delta_t.

Other suggestions are welcome.

**Comment by Expertium** [comment]
But the value of stability depends on what data is available. Do you want a two-step procedure?
1) Estimate S0 while removing outliers based on the current condition
2) Re-estimate S0 while removing outliers using two conditions, with the first estimate being used in the ratio
Like that?

**Comment by user1823** [comment]
Looks like you didn't notice this:

> Note: This suggestion is for trainset. I recommend keeping the pretrain filter unchanged.

I am suggesting to calculate S0 using the current filter and then using the two conditions to filter the reviews when training the other parameters.

**Comment by user1823** [comment]
> If a card is reviewed too early or too late, the response will be unreliable.

By the way, this statement holds true for any review, not just the first review. But, because filtering the outliers in each review is very difficult, we are filtering the outliers at the first review only.

(First review = second rating)

**Comment by Expertium** [comment]
> I am suggesting to calculate S0 using the current filter and then using the two conditions to filter the reviews when training the other parameters.

Ah, ok, I misunderstood.

**Comment by Expertium** [comment]
@L-M-Sherlock I suggest trying user1823's idea: https://github.com/open-spaced-repetition/fsrs-rs/issues/121#issuecomment-1836983587

**Comment by user1823** [comment]
> I suggest trying user1823's idea

In case you guys have forgotten, I want to remind you of the following:

In general, any change that makes the data more heterogeneous (such as the above proposed change) will tend to increase the RMSE. So, we shouldn't rely on RMSE to determine the effectiveness of this change.

It is possible that RMSE can decrease after this change, but it will happen only if the idea is VERY good. This happened in https://github.com/open-spaced-repetition/fsrs-optimizer/issues/16

In my opinion, the condition for implementing this idea should be that the change should not significantly increase the RMSE.

**Comment by user1823** [comment]
@L-M-Sherlock, what is your opinion on this suggestion?

If you are busy, it is fine. But you can at least give your opinion.

**Comment by L-M-Sherlock** [comment]
> But you can at least give your opinion.

I'm open for that. But the real world is more complicate than my expectation:

![image](https://github.com/open-spaced-repetition/fsrs-rs/assets/32575846/d58baac5-7a18-4d35-a3cc-9eb3bb20b3bb)

**Comment by user1823** [comment]
Maybe also filter reviews of cards with initial delta_t > 7 × S irrespective of whether they are filtered by the pretrain filter or not?

(delta_t = 7S implies R = 56.3%)

**Comment by user1823** [comment]
> I'm open for that. But the real world is more complicate than my expectation:

@L-M-Sherlock, the issue that you highlighted is not caused by my suggestion; it exists in the current version too. So, it should not be the reason to not apply my suggestion.

Basically, the reason behind my suggestion is that higher the amount of data available, better would be the trained weights (provided the data is not completely rubbish).

So, just because there are few cards for a particular delta_t, we should not discard all the ratings for those cards. The criteria to discard them should be stricter (as proposed above).

Similarly, just because a card falls in a delta_t with R = 100% (https://github.com/open-spaced-repetition/fsrs-rs/pull/135), we shouldn't discard all the ratings for those cards.

For sure, the above two cases are not suitable for calculating S0, but their subsequent ratings can be useful.

**Comment by Expertium** [comment]
> For sure, the above two cases are not suitable for calculating S0

The second case is suitable though. We use additive smoothing, so R will never be exactly 0% or 100% in the pretrain data. I haven't looked at the Rust code, but I'm assuming additive smoothing is implemented. If not, it definitely should be.

**Comment by L-M-Sherlock** [comment]
> Wait, Sherlock, then what's the point of removing cards with R=100%?

It's really weird that the retention is 100% when the interval >= 1 day. Removing these data could increase the pertaining's robustness.

**Comment by Expertium** [comment]
Is this issue still necessary?

**Comment by user1823** [comment]
Yes, the previous few changes were related to the pretrainset. This issue is related to the trainset.

**Comment by L-M-Sherlock** [comment]
I want to keep the current outlier filter in card level. All reviews of the same cards are filtered out if the first review is filtered out in the pretrainset.

**Comment by L-M-Sherlock** [comment]
> The pretrain and train methods are completely different.

Their goals are the same: minimizing the loss.

> Here, the optimizer is saying that the initial stability is 14 days but the optimizer completely ignores the cards whose first interval is more than 5 days, just because there are 1000s of reviews with first interval of 1 day (because of past Anki use).

The current outlier filter only filters out 5% of cards for each first rating. That's 1000 * 0.05 = 50 cards. When the number of cards whose first interval is more than 5 days exceeds 50, they will be taken into account.

**Comment by user1823** [comment]
> Their goals are the same: minimizing the loss.

Yes, their goals are the same. But, the method of achieving the goal is very different. Pretrain uses binning (based on delta_t) while train doesn't.

> When the number of cards whose first interval is more than 5 days exceeds 50, they will be taken into account.

In training, each card is treated independent of the others (because there is no binning). So, it shouldn't matter how many cards have a particular delta_t (except for some wierd cases like first review after 1000 days of graduating the card, which should be filtered out by the filter condition suggested by me). 

I agree that not a very large number of cards are affected. But, I can't think of any good reason for not including these cards in training.

**Comment by Expertium** [comment]
I don't have a strong opinion on this matter, but if I had to choose, I would choose to implement user1823's idea.

**Comment by user1823** [comment]
> If the number of reviews of the same delta_t is more than 6 in the pretrainset, the current outlier filter keeps them as well.

No

If the number is less than 6, then the reviews are filtered out. But if the number ≥ 6, then the decision of whether to keep them or not is based on whether they come in the bottom 5% of the total number of cards with that first rating or not.

This 5% can be quite large. For example, if I have 4000 cards with first rating of Good, then the optimizer is ignoring 200 cards for no good reason. This is not to say that the 5% limit should be adjusted. It is required for pretrain. But, when training, the filter should be slightly more loose.

This is my last comment on this issue. If you still don't want to implement it, then we can forget about it.
